yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -540,6 +565,28 @@ private[spark] class Client(
     amContainer
   }
 
+  def setupCredentials(): Unit = {
+    if (args.principal != null) {
+      if (args.keytab == null) {
+        throw new SparkException("Keytab must be specified when principal is specified.")
+      }
+      logInfo("Attempting to login to the Kerberos" +
+        s" using principal: ${args.principal} and keytab: ${args.keytab}")
+      val f = new File(args.keytab)
+      // Generate a file name that can be used for the keytab file, that does not conflict
+      // with any user file.
+      keytabFileName = f.getName + "-" + UUID.randomUUID().toString
```

Comment:
So, the semantics of this method are a little weird. There are many side-effects depending on what's being done: sometimes it sets `keytabFileName` /  `loginFromKeytab`, sometimes it sets `credentials`. Can this be cleaned up so that there's less state being passed around in these instance variables?